test/unit/logic/delegate.js

```diff
@@ -0,0 +1,535 @@
+				done();
+			});
+		});
+
+		it('should call callback with error if username is not lower case', function (done) {
+			trs.asset.delegate.username = 'UiOjKl';
+
+			delegate.verify(trs, sender, function (err) {
+				expect(err).to.equal('Username must be lowercase');
+				done();
+			});
+		});
+
+		it('should call callback with error if username is longer than 20 characters', function (done) {
+			trs.asset.delegate.username = Array.apply(null, Array(21)).map(function () {
+				return 'n';
+			}).join('');
+
+			delegate.verify(trs, sender, function (err) {
+				expect(err).to.equal('Username is too long. Maximum is 20 characters');
+				done();
+			});
+		});
+
+		it('should call callback with error if username is empty', function (done) {
+			trs.asset.delegate.username = '';
+
+			delegate.verify(trs, sender, function (err) {
+				// Cannot check specific error because '' coerces to false and we get error: Username is undefined
+				expect(err).to.exist;
+				done();
+			});
+		});
+
+		it('should call callback with error if username is address like', function (done) {
+			trs.asset.delegate.username = '163137396616706346l';
+
+			delegate.verify(trs, sender, function (err) {
+				expect(err).to.equal('Username can not be a potential address');
+				done();
+			});
+		});
+
+		it('should call callback with error when username contains symbols', function (done) {
+			trs.asset.delegate.username = '^%)';
+
+			delegate.verify(trs, sender, function (err) {
+				expect(err).to.equal('Username can only contain alphanumeric characters with the exception of !@$&_.');
+				done();
+			});
+		});
+
+		it('should be okay when username contains symbols which are valid', function (done) {
+			trs.asset.delegate.username = node.randomUsername() + '!@.';
+			accountsMock.getAccount.withArgs({username: trs.asset.delegate.username}, sinon.match.any).yields(null, null);
+
+			delegate.verify(trs, sender, function (err, returnedTrs) {
+				expect(err).to.not.exist;
+				expect(returnedTrs).to.equal(returnedTrs);
+				done();
+			});
+		});
+
+		it('should call callback with error when username already exists', function (done) {
+			var expectedError = 'Error: could not connect to server: Connection refused';
+			accountsMock.getAccount.withArgs({username: node.eAccount.delegateName}, sinon.match.any).yields(expectedError);
+
+			delegate.verify(trs, sender, function (err) {
+				expect(err).to.equal(expectedError);
+				accountsMock.getAccount.verify();
+				done();
+			});
+		});
+
+		it('should call callback with error when username already exists', function (done) {
+			accountsMock.getAccount.withArgs({username: node.eAccount.delegateName}, sinon.match.any).yields(null, node.eAccount);
+
+			delegate.verify(trs, sender, function (err) {
+				expect(err).to.equal('Username already exists');
+				accountsMock.getAccount.verify();
+				done();
+			});
+		});
+
+		it('should be okay for valid transaction', function (done) {
```

Comment:
what does it mean should be ok? I suggest to add a seperate describe(`when transaction is valid`) and to check each of assertion separately in it

Reply:
What do you mean by checking each assertion separately? Because Delegate.verify function only checks for assertions. And do not return anything in case of a valid transaction.

Reply:
```
describe('when transaction is valid')
   it('should call accounts.getAccount')
   it('should call accounts.getAccount with ....')
   it('should call a callback with error = null')
   it('should call a callback with transactions results')
describe('when transaction is invalid')
```

Reply:
You are right. Thanks!